ftl/python/builder.py

```diff
@@ -27,82 +28,132 @@
+        pip_env['PATH'] = venv_dir + "/bin" + ":" + os.environ['PATH']
+        return pip_env
 
+    def _setup_app_dir(self, app_dir):
         # Copy out the relevant package descriptors to a tempdir.
         for f in self.descriptor_files:
-            # if self._ctx.Contains(f):
-            with open(os.path.join(tmp_app, f), 'w') as w:
-                w.write(self._ctx.GetFile(f))
+            if self._ctx.Contains(f):
+                with open(os.path.join(app_dir, f), 'w') as w:
+                    w.write(self._ctx.GetFile(f))
 
-        tar_path = tempfile.mktemp()
+    def _setup_venv(self, dir_map, py_version):
         logging.info('Starting venv creation ...')
-
-        # TODO(aaron-prindle) add support for different python versions
         subprocess.check_call(
-            ['virtualenv', '--no-download', venv_dir, '-p', 'python3.6'],
-            cwd=tmp_app)
-        os.environ['VIRTUAL_ENV'] = venv_dir
-        os.environ['PATH'] = venv_dir + "/bin" + ":" + os.environ['PATH']
-        # bazel adds its own PYTHONPATH to the env
-        # which must be removed for the pip calls to work properly
-        my_env = os.environ.copy()
-        my_env.pop('PYTHONPATH', None)
+            ['virtualenv', '--no-download', dir_map['env'], '-p', py_version],
+            cwd=dir_map['app'])
+        logging.info('Finished venv creation')
 
+    def _pip_install(self, dir_map):
+        logging.info('Startin pip install')
         subprocess.check_call(
-            ['pip', 'install', '-r', 'requirements.txt'],
-            cwd=tmp_app,
-            env=my_env)
+            ['pip', 'wheel', '-w', dir_map['wheel'], '-r', 'requirements.txt'],
+            cwd=dir_map['app'],
+            env=self._gen_pip_env(dir_map['env']))
         logging.info('Finished pip install.')
 
-        logging.info('Starting to tar pip packages...')
-        subprocess.check_call(['tar', '-C', tmp_venv, '-cf', tar_path, '.'])
-        logging.info('Finished generating tarfile for pip packages...')
+    def _resolve_whls(self, dir_map):
+        return [
+            os.path.join(dir_map['wheel'], f)
+            for f in os.listdir(dir_map['wheel'])
+        ]
 
-        # We need the sha of the unzipped and zipped tarball.
-        # So for performance, tar, sha, zip, sha.
-        # We use gzip for performance instead of python's zip.
-        sha = 'sha256:' + hashlib.sha256(open(tar_path).read()).hexdigest()
-
-        logging.info('Starting to gzip pip package tarfile...')
-        subprocess.check_call(['gzip', tar_path])
-        logging.info('Finished generating gzip pip package tarfile.')
-        return open(os.path.join(tmp_venv, tar_path + '.gz'), 'rb').read(), sha
+    def _whl_to_fslayer(self, whl, venv_dir):
+        tmp_dir = tempfile.mkdtemp()
+        pkg_dir = os.path.join(tmp_dir, 'env')
+        os.makedirs(pkg_dir)
+        subprocess.check_call(
+            ['pip', 'install', '--prefix', pkg_dir, whl],
```

Comment:
Why do we need this second pip install? Can we re-use the one done during the initial pip wheel?

Reply:
I believe the first pip install only fetches the wheel, this one is exploding them